org.jacoco.core.test/src/org/jacoco/core/internal/analysis/filter/FinallyFilterTest.java

```diff
@@ -0,0 +1,420 @@
+/*******************************************************************************
+ * Copyright (c) 2009, 2017 Mountainminds GmbH & Co. KG and Contributors
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    Evgeny Mandrikov - initial API and implementation
+ *
+ *******************************************************************************/
+package org.jacoco.core.internal.analysis.filter;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.fail;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import org.jacoco.core.internal.instr.InstrSupport;
+import org.junit.Test;
+import org.objectweb.asm.Label;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.tree.AbstractInsnNode;
+import org.objectweb.asm.tree.MethodNode;
+
+public class FinallyFilterTest implements IFilterOutput {
+
+	private final IFilter filter = new FinallyFilter();
+
+	private final MethodNode m = new MethodNode(InstrSupport.ASM_API_VERSION, 0,
+			"name", "()V", null, null);
+
+	/**
+	 * <pre>
+	 *   try {
+	 *     ...
+	 *     if (...) {
+	 *       ...
+	 *       return;
+	 *     } else {
+	 *       ...
+	 *       return;
+	 *     }
+	 *   } finally {
+	 *     ...
+	 *   }
+	 * </pre>
+	 */
+	@Test
+	public void should_analyze_control_flow() {
```

Comment:
@Godin While I'm trying to wrap my mind around the filter implementation I was wondering why we need to loop over all tryCatchBlocks with the same handler. This seems to be the only test case which triggers this. Therefore I wonder whether it might make sense to add this scenario also to the validation test.

Reply:
@marchof done

Reply:
@Godin When I remove the two loops like this:

		// for (final TryCatchBlockNode t : tryCatchBlocks) {
		{
			final TryCatchBlockNode t = finallyBlock;

the validation tests are still fully green (ECJ) and a different test fails for javac:

    emptyCatch(org.jacoco.core.test.filter.FinallyTest): Status in line 79:	nop("finally"); // $line-emptyCatch.1$ expected:<[FUL]LY_COVERED> but was:<[PART]LY_COVERED>

But new test `twoRegions()` stays green. What am I missing here?

Reply:
@marchof after this modification of both loops
* `emptyCatch` fails because condition of `if` statement in second loop always `false`
* unit test for the case of `twoRegions` shows that merged too much, but validation test has assertion only for `final`, for same reason `ecj` stays green

Missing assertions added.

Also this test demonstrates https://bugs.openjdk.java.net/browse/JDK-7008643